cm/cm_hdd_cdd/hddcdd.py

```diff
@@ -0,0 +1,214 @@
+TESTDATA_DIR = CURRENT_FILE_DIR / "testdata"
+
+
+def scale_stat(stats: dict, factor):
+    """From a list of stats, take the factor into account and
+    modify the stats accordingly.
+    """
+    for stat_name, stat in stats.items():
+        for scaled_stat_prefix in SCALED_STATS_PREFIX:
+            if stat and stat_name.startswith(scaled_stat_prefix):
+                stats[stat_name] = stat * factor
+
+
+def compute_centroid(geojson: Dict) -> Tuple[float, float]:
+    """
+    Example
+    >>> gj = {'features': [
+    ...          {'geometry':
+    ...               {'coordinates': [[[11.061588, 45.567844],
+    ...                                 [11.055015, 45.563899],
+    ...                                 [11.050421, 45.565349],
+    ...                                 [11.040453, 45.560869],
+    ...                                 [11.032734, 45.561389]]],
+    ...                'type': 'Polygon'},
+    ...           'id': 80284,
+    ...           'properties': {'ds_id': 0,
+    ...                          'dt': '',
+    ...                          'fields': '',
+    ...                          'id': 'IT_023038',
+    ...                          'layer': '{ "type": "numerical" }',
+    ...                          'start_at': '',
+    ...                          'units': '{ "": null }',
+    ...                          'variables': '{ "": 0 }',
+    ...                          'z': ''},
+    ...           'type': 'Feature'}],
+    ...       'type': 'FeatureCollection'}
+    >>> compute_centroid(gj)
+    (11.048, 45.564)
+    """
+    try:
+        coords = np.array(geojson["features"][0]["geometry"]["coordinates"])
+    except KeyError:
+        logging.error(geojson)
+        raise ValueError(
+            "FAILED! The provided geometry is not a correct/supported geojson format."
+        )
+    return tuple(np.around(coords[0].mean(0), decimals=DECIMALS))
+
+
+def get_datarepodir() -> Path:
+    return Path(os.environ.get("CM_HDD_CDD_REPOSITORY", TESTDATA_DIR.as_posix()))
+
+
+@lru_cache(maxsize=256)
+def get_datadir(
+    datarepository: Path,
+    sim_type: str = "historical",
+    dd_type: str = "hdd",
+    Tb=18.0,
+    aggr_window="monthly",
+    method="average",
+) -> Path:
+    """
+    >>> get_datadir("cm/cm_hdd_cdd/testdata").as_posix()
+    "cm/cm_hdd_cdd/testdata/historical/hdd/18.0/monthly/average/"
+    """
+    return (
+        Path(datarepository) / sim_type / dd_type / f"{Tb:.1f}" / aggr_window / method
+    )
+
+
+@lru_cache()
+def extract_by_dir(
+    gdir: Path,
+    lon: float,
+    lat: float,
+    __datasets: Dict[str, rasterio.DatasetReader] = {},
+):
+    res = []
+    idx = []
+    for gfi in gdir.iterdir():
+        if gfi.name.endswith(".tif"):
+            idx.append(gfi.name[:-4])
+            try:
+                gx = __datasets[gfi.as_posix()]
+            except KeyError:
+                gx = rasterio.open(gfi)
+                __datasets[gfi.as_posix()] = gx
```

Comment:
I cannot close the file, the local variable: `__datasets: Dict[str, rasterio.DatasetReader] = {}` is creating a cache with `rasterio.DatasetReader` instances, to avoid to open the file every time the server is receiving a request from a client. If I close the rasterio instance than I would not be able to query the raster later using the `gx` variable:

```python
@lru_cache()
def extract_by_dir(
    gdir: Path,
    lon: float,
    lat: float,
    __datasets: Dict[str, rasterio.DatasetReader] = {},
):
    res = []
    idx = []
    for gfi in gdir.iterdir():
        if gfi.name.endswith(".tif"):
            idx.append(gfi.name[:-4])
            try:
                gx = __datasets[gfi.as_posix()]
            except KeyError:
                gx = rasterio.open(gfi)
                __datasets[gfi.as_posix()] = gx
            yp, xp = gx.index(lon, lat)      # <=here I'm using the raterio gx instance
            val = gx.read()[0][yp, xp]       # <=here I'm using the raterio gx instance
            logging.info(
                f"{gfi.name}@{lon:.{DECIMALS}f},{lat:.{DECIMALS}f}: {yp}, {xp} => {val}"
            )
            res.append(val)
    sr = pd.Series(np.array(res), index=idx, name=f"yp={yp},xp={xp}")
    sr.sort_index(inplace=True)
    return sr
```
Open the file without closing it shouldn't be a big issue here, because:
* we are opening the file in read-only mode;
* we are doing so only once per server process not per client request

We might need to raise the docker open file limits,  the default value in ubuntu is: 176772.